### ISSUE TITLE
Direct vs Indirect

### ISSUE THREAD

**Issue by kalmarek** [cvxgrp/scs#81]
The Indirect solver (on my problems) is consistently slower (by large margin) than the direct; Any idea why this may happen? (both are multithreaded @4 cores), and the problems are relatively large:
e.g. `n = 1959211, constraints m = 5937302`
100 iterations takes 1.56e+03 s for the direct and 1.98e+03 for the indirect;

Any idea why these may happen?

I use these through SCS.jl (https://github.com/kalmarek/SCS.jl/tree/indirect)

**Comment by bodono** [comment]
Firstly, how many CG iterations on average does the indirect method need? It prints it out at the very end:
```
Lin-sys: avg # CG iterations: 4.23, avg solve time: 9.70e-04s
```

The indirect method parallelizes over columns for the multiply and over rows for the transpose multiply, is your data heavily concentrated in a single row / column? If so that thread would be dominating.

One test you can do is to compile with the environment variable EXTRA_VERBOSE=1, (e.g. `make EXTRA_VERBOSE=1` in bash). Running an example after that will dump a lot of info, but if you look for lines like:
```
mult By A trans time: 2.78e-05s
mult By A trans time: 3.74e-05s
```
It will tell you how long your multiplies are taking. If every second one is taking a really long time, then it means either your multiply or transpose multiply is dominating (both the multiply and the transpose multiply take the transpose code path for efficiency).

**Comment by kalmarek** [comment]
hmm I don't quite about the concentration You ask (sorry, I'm very new to the field):
In the formulation which I specify the problem i.e.
```
min <C,X>
subject to <A_k,X> = b_k
                            X ≥ 0
```
`X` is of size ~`2000×2000`, `k = 2018880`, and `A_k`s are extremely sparse (A_1) is diagonal, only (the first) 2000 `A_k`s contain ~100 nnz; the remaining, only a few... Is this the concentration You meant?
tail of the log:
```
----------------------------------------------------------------------------
Status: Solved/Inaccurate
Hit max_iters, solution may be inaccurate
Timing: Solve time: 2.01e+05s
	Lin-sys: avg # CG iterations: 8.01, avg solve time: 1.11e+00s
	Cones: avg projection time: 8.83e+00s
----------------------------------------------------------------------------
```

**Comment by bodono** [comment]
In that log the cone projection is the dominant cost, so perhaps there is not much to be gained using the indirect solver? Each iterations consists of one lin-system solve and one cone projection, so if it takes 8x the time to do the cone projection than the lin-sys solve then that step will dominate.

In SCS there is a data matrix `A`, which consists of your `A_k` matrices, and perhaps some other stuff depending on how it was built. If this matrix matrix has a lot of non-zeros in one column (or one row), then the matrix multiply will be dominated by that one column (or row), since each thread takes one column (row) in the multiply. For example if each column has 10 non-zeros, except for one which has 10k non-zeros, then the cost of the matrix multiply will mostly be the cost of doing the multiply with that one column and so the multi-threading won't get you much.

**Comment by kalmarek** [comment]
Thanks for the answer; Sure, the dominating factor is the cone projection step, but still I expected indirect solver to be faster than the direct one; 

I gathered a few timings on the problem (with `-DEXTRAVERBOSE=1`), before running out of space ;P
```
mult By A time: 6.58e-02s
mult By A trans time: 6.02e-02s
mult By A time: 3.51e-02s
mult By A trans time: 4.39e-02s
mult By A time: 6.69e-02s
mult By A trans time: 4.16e-02s
mult By A time: 5.50e-02s
mult By A trans time: 5.20e-02s
mult By A time: 6.65e-02s
mult By A trans time: 4.23e-02s
mult By A time: 6.06e-02s
mult By A trans time: 8.60e-02s
mult By A time: 3.56e-02s
mult By A trans time: 3.37e-02s
mult By A time: 4.60e-02s
mult By A trans time: 3.22e-02s
```
so they look pretty evenly matched;
and most importantly:
```
indirect:
Timing: Solve time: 1.74e+04s
	Lin-sys: avg # CG iterations: 7.96, avg solve time: 1.06e+00s
	Cones: avg projection time: 7.51e+00s
```
```
direct:
Timing: Solve time: 1.53e+04s
	Lin-sys: nnz in L factor: 15659737, avg solve time: 1.58e-01s
	Cones: avg projection time: 7.38e+00s
```
So it seems that on this particular class of problems the direct Lin-sys is order of magnitude faster than the CG

**Comment by bodono** [comment]
The indirect solver only uses multiplies by A transpose, in order to parallelize more efficiently, so that printout must be from the direct solver? Try the indirect one and you might see a difference (they will all be `mult By A trans time`, you shouldn't see any `mult By A time`). Also be sure you are using multiple threads to give the indirect solver a fair showing. Just seeing the number of threads in use higher than 1 is not sufficient, because BLAS / LAPACK can be multi-threaded when computing the projection onto the semi-definite cone. To really test you would probably have to try an example without SD cones.

The direct solver solves are loosely speaking `O(n^2)`, the indirect solver uses matrix multiples so for small CG iterations each solve is also `O(n^2)`, probably with worse constants. The advantage that the indirect solver has are:
1) It doesn't do a factorization at the start, which the direct solver does and is very expensive at `O(n^3)`, you can see how long that took in the header `Setup time: 2.63e-01s`.
2) The direct solver might find a bad factorization as judged by fill-in factor, you can calculate this by dividing `nnz in L factor: 519024` by `nnz in A = 32000` (in this case it's > 16, which is very bad, but expected since I am testing a totally random A matrix, in this case the direct solver is slower per-solve than the indirect solver running on a single thread, that's how bad it is).
3) It can be multi-threaded easily.

So in short it might not always be faster, but often will.

**Comment by kalmarek** [comment]
Sorry, my bad that was the direct indeed; below is the indirect;
Yeah, so far I thought that cpu utilisation graph indicates that things are computed in parallel, but now I am not sure;-), it is possible that only projection is done in multiple threads, as I can see plateaus of high cpu-load and short spikes of single-threaded use, which may correspond to the cone projection taking the bulk of time...

So, to be absolutely clear: the `-DOPENMP`, `-fopenmp` and `-lgomp` flags I can provide during the compilation/linking only influence parralelisation (or lack thereof) lin-sys phase of the indirect solver??

Setup time of the indirect is shorter: ~0.5s vs ~5.5s of the direct;

The fill-in factor for my problems is slightly less than 2, so this is not a factor;

I'll get back, when the new timings arrive

```
mult By A trans time: 3.97e-02s
mult By A trans time: 3.71e-02s
mult By A trans time: 3.59e-02s
mult By A trans time: 4.66e-02s
mult By A trans time: 4.81e-02s
mult By A trans time: 5.18e-02s
mult By A trans time: 5.94e-02s
mult By A trans time: 5.93e-02s
mult By A trans time: 3.87e-02s
mult By A trans time: 3.69e-02s
mult By A trans time: 3.53e-02s
mult By A trans time: 3.99e-02s
mult By A trans time: 3.10e-02s
mult By A trans time: 4.60e-02s
mult By A trans time: 4.31e-02s
mult By A trans time: 4.35e-02s
mult By A trans time: 3.39e-02s
mult By A trans time: 5.12e-02s
mult By A trans time: 4.32e-02s
mult By A trans time: 7.23e-02s
mult By A trans time: 4.37e-02s
mult By A trans time: 5.34e-02s
mult By A trans time: 3.38e-02s
mult By A trans time: 3.71e-02s
mult By A trans time: 3.36e-02s
mult By A trans time: 4.04e-02s
mult By A trans time: 4.05e-02s
mult By A trans time: 3.70e-02s
mult By A trans time: 3.16e-02s
mult By A trans time: 3.70e-02s
mult By A trans time: 3.30e-02s
mult By A trans time: 3.78e-02s
mult By A trans time: 3.50e-02s
mult By A trans time: 4.48e-02s
mult By A trans time: 4.92e-02s
mult By A trans time: 3.90e-02s
mult By A trans time: 4.06e-02s
mult By A trans time: 4.72e-02s
mult By A trans time: 3.63e-02s
mult By A trans time: 3.75e-02s
mult By A trans time: 3.15e-02s
mult By A trans time: 3.47e-02s
mult By A trans time: 3.18e-02s
mult By A trans time: 3.63e-02s
mult By A trans time: 3.17e-02s
mult By A trans time: 3.87e-02s
mult By A trans time: 3.26e-02s
mult By A trans time: 4.05e-02s
mult By A trans time: 3.54e-02s
mult By A trans time: 5.13e-02s
mult By A trans time: 3.86e-02s
mult By A trans time: 4.88e-02s
```

**Comment by bodono** [comment]
From those timings it doesn't look like either the forward or transpose multiply is really dominating, which is good.

As well as those flags you might need an environment variable set, `export OMP_NUM_THREADS=16` for example (for 16 threads in that case). You don't need `-DOPENMP`, that flag has been removed, but it is assuming that the compiler is setting `_OPENMP` to true (which should happen when using `-fopenmp`). You could try setting that manually and see if it makes a difference.

Since the fill-in factor is low, and the cost of the direct factorize step is only a few seconds, this might be a case where the direct solver is indeed slightly faster.

**Comment by kalmarek** [comment]
Timings arrived:
```
Timing: Solve time: 1.68e+04s
	Lin-sys: avg # CG iterations: 7.96, avg solve time: 8.56e-01s
	Cones: avg projection time: 7.41e+00s
```
So, the avg solv time did decreased but not by much: 1.05 vs 0.86 s (this is run with `OMP_NUM_THREADS=4` on four real cores); It seems that the direct solver is still faster for this problem.
Thanks for settling the matter concerning parallelisation!
